sdlib/d/gc/tcache.d

```diff
@@ -806,3 +859,33 @@ unittest arraySpill {
 	testSpill(16384, [0, 1, 2, 500, 16000, 16383, 16384]);
 	testSpill(20480, [0, 1, 2, 500, 20000, 20479, 20480]);
 }
+
+unittest finalization {
+	// Faux destructor which simply records most recent kill:
+	static size_t lastKilledUsedCap = 0;
+	static void* lastKilledPtr;
+	static void destruct(void* body, size_t size) {
+		lastKilledUsedCap = size;
+		lastKilledPtr = body;
+	}
+
+	// Finalizers for large allocs:
+	auto s0 = threadCache.allocAppendable(16384, false);
+	threadCache.setFinalizer(s0, &destruct);
+	threadCache.destroy(s0);
+	assert(lastKilledPtr == s0);
+	assert(lastKilledUsedCap == 16384);
+
+	// Resize preserves finalizer:
+	auto s1 = threadCache.allocAppendable(20000, false);
+	threadCache.setFinalizer(s1, &destruct);
+	auto deadspace = threadCache.allocAppendable(20000, false);
+	auto s2 = threadCache.realloc(s1, 60000, false);
+	assert(s2 !is s1);
+	threadCache.destroy(s2);
+	assert(lastKilledPtr == s2);
+	assert(lastKilledUsedCap == 60000);
+
+	// Destroy on non-finalized alloc is harmless:
+	threadCache.destroy(deadspace);
```

Comment:
Make sure nothing was destroyed. For instance, you could have a destroy count, and check it did not move.